src/Entities/Player.cpp

```diff
@@ -1231,6 +1231,21 @@ void cPlayer::SetGameMode(eGameMode a_GameMode)
 void cPlayer::LoginSetGameMode( eGameMode a_GameMode)
 {
 	m_GameMode = a_GameMode;
+	
+	if (!(IsGameModeCreative() || IsGameModeSpectator()))
+	{
+		SetFlying(false);
+		SetCanFly(false);
+	}
+
+	if (IsGameModeSpectator() && IsVisible())
```

Comment:
I'm not sure if this is how spectator works, but if a player is in spectator mode and they aren't visible (normal presumably?), does that mean it will fall through to the else if and set them to be visible?